### DISCUSSION TITLE
Few Questions

### DISCUSSION THREAD

**Discussion by cloudcompute** [ricsanfre/pi-cluster#127] [Q&A] [answered]
Hi

This project seems promising. I have few questions, could you please answer:

a. Is it possible to deploy the `both master and worker nodes on a laptop` having 16 GB and 4 cores?

You have a very good knowledge about the K8S ecosystem, great if you answer these two..

b. Would you recommend using a tool like skaffold, devspace, garden, or telepresence. The point is to
    i . Skip building the container images and witness the changes made to the code instantly i.e. `hot reloading`.
    ii. `Debugging`

c. `Traefik` is an open source reverse proxy. But it `lacks certain features` which are available in its Enterprise version. In this context, would you `recommend using Traefik` or any other alternative? `How about Istio`, it can be used as API gateway and Ingress both.

Thanks in anticipation
Raman

**Comment by ricsanfre** [comment, marked as answer]
Hi Raman, 

a) This project goal is to build a kubernetes homelab using baremetal low cost nodes (raspberry pi). You can try to build your own cluster for learning purposes using a single hardware box and aplying virtualization to have several virtual nodes.
   As an alternative you can use dockerized version of K3S: [k3d](https://k3d.io). K3d runs on a single box and it can be used as development or training environment. In this case, you will be able to test most of the services but not all, for example: longhorn is not supported in k3d. https://k3d.io/v5.3.0/faq/faq/#longhorn-in-k3d.
  In both cases, parts of this project can be reused, but it will require adaptation to the virtualized environment docker or VMs.
  Single box with 16GB and 4 cores could not be enough for deploying all services but at least you  can use it as playground to learn or test.

b) I am afraid I do not have any experience on such tools. Cannot provide any insights on that.

c) I am using Traefik's Ingress Controller, not its API gateway. So far, for me, open source edition is enough for my homelab use cases. What are the features you are missing?
Regarding Istio. Istio is mainly service mesh solution. For service mesh, I am using linkerd which is easier to install and configure than Istio (moreover Istio support of arm64 architectures has been recently added). Traefik can be also easily integrated with linkerd.
Istio is implementing k8s gateway api, a evolution of the Ingress Controller to support not only HTTP/HTTPS protocol. Traefik opensource has also its own implementation. This kind of gateway is different from the concept of API gateway in a microservices architecture. For that other opensource projects can be used: Tyk, Kong, etc.


Regards
Ricardo

**Comment by cloudcompute** [comment]
Hello Ricardo

Thank you for your detailed response. I will go through k3d. I executed the command 'nproc' and the output was 8. So it is probably a 8 core CPU.

[Here](https://traefik.io/pricing/) is the difference between the Traefik's open-source and enterprise versions. The main feature is the lack of support for Security which is a must to ensure authentication and authorization. Another feature missing is high availability and scalability.

[This article](https://tetrate.io/blog/istio-service-mesh-api-gateway/) indicates that we can use Istio as an API Gateway. But again, it seems that few features are missing and Tetrate is trying to sell its product that makes Istio a full-fledged API gateway.
 
I want to avoid using semi open-source software. The only reason is it forces us to continue using their products.. vendor lock-in.

I took a closer look at Kong.. again lock-in.

With Regards
Raman
